Attempt to fix issue 11719

AthulJoshi opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

This pull request is to try and implement #11719 

I went through the docs, and the package and came to a conclusion.

- If the frame of the SkyCoord object being passed is "galactic", then the center object would have "l and b" as the data members, and for the other three frames(I read that only 4 frames are preferred while using SpericalCircle plot), the center object would have "ra and dec" as the data members.

- And also that SphericalCircle should accept "center" as a tuple, the current implementation.

- I don't know if importing SkyCoord is a good idea, I'll look for an alternate method.

I've implemented this approach, not sure if this is the right way, please correct me if I'm wrong.

AthulJoshi commented:
If the implementation is good, I will proceed to add the required docs. And I guess I'll have to modify the tests too right?

hamogu commented:
I think this look great. Thank you for contributing to astropy! I need to think a little about the "galactic". 
Importing SkyCoord is fine. We are in astropy, and we can use all astropy modules - that's why this is one package. And you are right, you'll have to change the docstring and add a test. I would also shows this off in the last example on this page : https://docs.astropy.org/en/latest/visualization/wcsaxes/overlays.html#spherical-patches (code here: https://github.com/astropy/astropy/blob/main/docs/visualization/wcsaxes/overlays.rst). It might be the easiest to add a second circle to the same image that uses a Skycoord. Then, you build the docs locally (let me know if you need help) and look at it and you will see immediately if it all works.

You'll see that all this extra stuff (doctring, test, etc.) often takes far more time to get it right then the code itself. That is normal! It might be frustrating, but it's the only method to make sure that astropy really is this high quality software package that people like.

And one small hint:
If you open a PR that solves an issue you can put something like this in the commit message on git:
`This closes #11719.` or `This fixes #11719.` If you use any of the words "close, closes, closed, fixes, fixed" followed by a "#" and a number, github will automatically close the issue when this PR is merged. It's not really required (we cal always close that by hand), I just thought I'll let you know ;-)

AthulJoshi commented:
@hamogu, I've added the feature to the docs. I have a question. I looked into the codebase and couldn't find any tests specific to SphericalCircle. So,  I'm not sure where to add the test.

hamogu commented:
I searched for "sphericalcircle" in the github search bar on the top left: https://github.com/astropy/astropy/search?q=sphericalcircle
The last file listed has a test. That test makes and image and compares the png pixel-by-pixel with a reference image. Thus, testing images is a little more complicated than testing other code: https://docs.astropy.org/en/latest/development/testguide.html#image-tests-with-pytest-mpl
I think we can do something simple here that saves us a lot of work. Simply *draw the exact same circle* on top of the one that the test already draws, so that the output image should be the same. If it's not, something is wrong!
So, in this line :https://github.com/astropy/astropy/blob/df3e5bb36120797065c8e326309663ab1a3df788/astropy/visualization/wcsaxes/tests/test_images.py#L667

You can change it to something very similar to this (I did not test that, there might be a typo somewhere and I did not get the line breaks right):
```
r1 = SphericalCircle((266.4 * u.deg, -29.1 * u.deg), 0.15 * u.degree, edgecolor='purple', facecolor='none', transform=ax.get_transform('fk5'))
ax.add_patch(r1)
# Now draw the same region again on top. Should give the same image.
r2 = SphericalCircle(SkyCoord(266.4 * u.deg, -29.1 * u.deg), 0.15 * u.degree, edgecolor='purple', facecolor='none', transform=ax.get_transform('fk5'))
ax.add_patch(r2)
```

This would then show that the two match pretty well. They could still be entirely broken (e.g. if your new code puts the circles 90 deg to the west, if would not be seen at all, and thus the image to compare would look the same. We need to test for that, too in some way. We can compare the matplotlib patches "r1" and "r2". Maybe `assert r1.get_center() == r2.get_center()` and get_radius() or something like that?

AthulJoshi commented:
Ooh, sorry about that. Yeah, I guess that would work. I'll go through the docs and try doing that. Thanks!

AthulJoshi commented:
@hamogu, It turns out I'm stuck at the testing part.
So, I tried changing this line:
[](https://github.com/astropy/astropy/blob/df3e5bb36120797065c8e326309663ab1a3df788/astropy/visualization/wcsaxes/tests/test_images.py#L667)
to:
`r1 = SphericalCircle(SkyCoord(266.4 * u.deg, -29.1 * u.deg), 0.15 * u.degree,
                             edgecolor='purple', facecolor='none',
                             transform=ax.get_transform('fk5'))
        ax.add_patch(r1)

        r2 = SphericalCircle((266.4 * u.deg, -29.1 * u.deg), 0.15 * u.degree,
                             edgecolor='purple', facecolor='none',
                             transform=ax.get_transform('fk5'))
        ax.add_patch(r2)`

But then, it gives an error :
x-special/nautilus-clipboard
copy
file:///home/athul/Pictures/test_images%20(copy).png

Then, I changed the line to
`        r = SphericalCircle(SkyCoord(266.4 * u.deg, -29.1 * u.deg), 0.15 * u.degree,
                             edgecolor='purple', facecolor='none',
                             transform=ax.get_transform('fk5'))
        ax.add_patch(r)
`
the test passes.

So, in short, when I place the circles on top of each other, the test fails. But when I changed the first circle to take a SkyCoord object as input, the test passes. The implementation is working but I can't figure out what is wrong.

And I found that 
`assert (r1.get_xy() == r2.get_xy()).all()`
could be used to solve the second problem.

AthulJoshi commented:
@hamogu, It turns out I'm stuck at the testing part.

So, I tried changing

https://github.com/astropy/astropy/blob/df3e5bb36120797065c8e326309663ab1a3df788/astropy/visualization/wcsaxes/tests/test_images.py#L667-L670

to:
https://gist.github.com/AthulJoshi/a1b6f4dcbc4ceb645cf0bcdf02d51f88

But then, it gives an error :
![test_images](https://user-images.githubusercontent.com/60695862/120163509-35591900-c217-11eb-9160-b7ad054d4336.png)

Then, I changed the line to



        r = SphericalCircle(SkyCoord(266.4 * u.deg, -29.1 * u.deg), 0.15 * u.degree,
                             edgecolor='purple', facecolor='none',
                             transform=ax.get_transform('fk5'))
        ax.add_patch(r)

and the test passes.

- So, in short, when I place the circles on top of each other, the test fails. But when I changed the only circle to take a SkyCoord object as input, the test passes. The implementation is working but I can't figure out what is wrong.

- And I found that `assert (r1.get_xy() == r2.get_xy()).all()` could be used to solve the second problem. Because, if all the coordinates used to plot the figures correspond with each other, the center and radius would be same right?

hamogu commented:
If `assert (r1.get_xy() == r2.get_xy()).all()` works, that's great. I was not sure if they do, but I guess they should. Have you looked at the two images locally to compare them?
Also, I think it's probably time to ask one of the maintainers of this sub-package to have a look. They know this code better than I do and may have further suggestions - I'll ping them: @astrofrog @larrybradley @Cadair

hamogu commented:
We are in the business of making code work, not to debug tests. I think the `assert` statement should do and looking for the other difference is a waste of time. So, let's put in the assert test and see if one of the sub-package reviewers has time to look at this.

AthulJoshi commented:
> If `assert (r1.get_xy() == r2.get_xy()).all()` works, that's great. I was not sure if they do, but I guess they should. Have you looked at the two images locally to compare them?
> Also, I think it's probably time to ask one of the maintainers of this sub-package to have a look. They know this code better than I do and may have further suggestions - I'll ping them: @astrofrog @larrybradley @Cadair

Thanks! Yep, I checked them locally, and both look similar to my eyes (^_^").

AthulJoshi commented:
> We are in the business of making code work, not to debug tests. I think the `assert` statement should do and looking for the other difference is a waste of time. So, let's put in the assert test and see if one of the sub-package reviewers has time to look at this.

Alright, I'll push the changes then.

hamogu commented:
@AthulJoshi Just to let you know what the process is. I think this looks good, so I'd like to wait  little longer if one of the maintainers for this sub-package has a chance to chime in. They know the ins and outs of the WCS visualization code better than I do. However, if they are all too busy and don't have the time to take a look and I'll take your code and try it on a few of my plots and, if I don't find anything else, I could merge it, too. So, now we need a little patience. Almost everyone working on astropy is a volunteer and sometimes we are on vacation or have urgent works things to do. However, if you don't hear anything for another week or so, please ping me to remind me to look at this again. We don't want to forget about this PR, it's too useful!

hamogu commented:
Of course, the tests should pass first! Let me know if you need any help with figuring this out.

AthulJoshi commented:
> @AthulJoshi Just to let you know what the process is. I think this looks good, so I'd like to wait little longer if one of the maintainers for this sub-package has a chance to chime in. They know the ins and outs of the WCS visualization code better than I do. However, if they are all too busy and don't have the time to take a look and I'll take your code and try it on a few of my plots and, if I don't find anything else, I could merge it, too. So, now we need a little patience. Almost everyone working on astropy is a volunteer and sometimes we are on vacation or have urgent works things to do. However, if you don't hear anything for another week or so, please ping me to remind me to look at this again. We don't want to forget about this PR, it's too useful!

Thanks! Actually, I've been waiting for someone to look into it. I know how OpenSource works, so no probs. But I was terrified by the number of unmerged PRs tho lol. I was actually looking for other issues that I could get into. And I have no idea how to fix the test, I'll do more reading on that part. I'll ping you in a week, if there's no activity over here.

hamogu commented:
For the failing test, I think you can go to keep the old test to compare the image and then you do
```
r2 = SphericalCircle(SkyCoord(266.4 * u.deg, -29.1 * u.deg), 0.15 * u.degree,
                            edgecolor='purple', facecolor='none',
                            transform=ax.get_transform('fk5'))
assert np.all(r2.get_xy() == r.get_xy())
```

AthulJoshi commented:
> For the failing test, I think you can go to keep the old test to compare the image and then you do
> 
> ```
> r2 = SphericalCircle(SkyCoord(266.4 * u.deg, -29.1 * u.deg), 0.15 * u.degree,
>                             edgecolor='purple', facecolor='none',
>                             transform=ax.get_transform('fk5'))
> assert np.all(r2.get_xy() == r.get_xy())
> ```

That would stop the tests from failing. I'll do it if it's okay, ill make the changes and push then.

AthulJoshi commented:
@hamogu , this good now?

hamogu commented on docs/visualization/wcsaxes/overlays.rst:
```diff
@@ -326,9 +326,20 @@ In the case where you are making a plot of a celestial image, and want to plot a
     ax.add_patch(r)
+
+    #The following lines show the usage of a SkyCoord object as the input.
+    skycoord_object = SkyCoord(266.4 * u.deg, -28.7 * u.deg)
```
Last comment from my side, I promise. For the test, we put the circles on top of each other of that we can do the assert statement. For the docs, won't it be better to have different circles so that both can be seen? Or at least a comment in the code something like: This command makes the same circle so you won't see it in the plot because they are on top of each other"?

AthulJoshi replied:
The circles in the docs can be seen seperately. Are you suggesting to have them on top of each other?

hamogu replied:
Ooh, OK, I see. Sorry.

hamogu commented on astropy/visualization/wcsaxes/patches.py:
```diff
@@ -81,9 +83,15 @@ class SphericalCircle(Polygon):
+        # from a tuple of two quantities or a single 2-element Quantity.
+        if isinstance(center, SkyCoord):
+            if center.frame.name == 'galactic':
+                longitude, latitude = center.l, center.b
```
OK, now, really last thing: I started the tests again and this line is not covered. Can you add another test? I don't think you need to draw anything for the test, so that we don't have to add an image. Just make any example locally so that you see it works and then you just test a single point similar to "assert np.allclose((r1.get_xy()[0], [123, 456])" where you obviously need to replace the 123 and 456 with whatever  number you get in the example.

You'll see that I approved it already, so it could be merged right now, but that little addition (just to project us form something stupid, say, accidentially switching l and b) would make it even better. Thank you so much for your patiance!

AthulJoshi replied:
I'll try to add that :)

eerovaher commented on astropy/visualization/wcsaxes/patches.py:
```diff
@@ -81,9 +83,15 @@ class SphericalCircle(Polygon):
+        # from a tuple of two quantities or a single 2-element Quantity.
+        if isinstance(center, SkyCoord):
+            if center.frame.name == 'galactic':
+                longitude, latitude = center.l, center.b
```
Wouldn't it be better to use `center.representation_component_names` to find out what the longitude and latitude of `center` are called instead of hard-coding the names? It would also allow the user to provide a `SkyCoord` in e.g. ecliptic coordinates, which is not possible in this implementation.

eerovaher commented on astropy/visualization/wcsaxes/patches.py:
```diff
@@ -81,9 +83,15 @@ class SphericalCircle(Polygon):
-        longitude, latitude = center
+        # Extract longitude/latitude, either from a SkyCoord object, or
+        # from a tuple of two quantities or a single 2-element Quantity.
+        if isinstance(center, SkyCoord):
```
It is possible to define a `SkyCoord` using Cartesian coordinates. In that case an attempt to access longitude and latitude will not succeed because those are coordinates in a spherical system. It would therefore be better to include a check if `center.representation_type` is `SphericalRepresentation`.

AthulJoshi replied:
@eerovaher Thanks for the review! I have a doubt that should an error be raised if a SkyCoord using Cartesian coordinates is passed.

eerovaher replied:
If the `representation_type` of the `SkyCoord` is not `SphericalRepresentation` then maybe the best thing to do would be to convert the `SkyCoord` into spherical coordinates but also to raise a warning to make sure that the user knows what is happening.

hamogu reviewed (changes requested):
I'm glad that someone with a deeper knowledge of the Coordinates packages has weighed in. I was worried about the l/b vs Ra/dec, but I could not point to the actual problem. I think the suggestions in the comment look reasonable.

AthulJoshi commented:
Ill try to make all the changes and push ASAP.

AthulJoshi commented:
@hamogu @eerovaher I've modified the code. But I didn't solve the coverage problem I guess, I tried to add an assert as @hamogu had suggested, but the local codecov report doesn't show any improvement. Maybe I didn't get what you had suggested.

eerovaher commented on astropy/visualization/wcsaxes/patches.py:
```diff
@@ -81,9 +84,20 @@ class SphericalCircle(Polygon):
+                warnings.warn(f'Received center of representation type {rep_type} '
+                              'will be converted to Spherical Representation ',
+                              UserWarning)
+                center.representation_type = 'spherical'
```
This has the side effect that the `SkyCoord` the user has provided is converted to spherical coordinates. We do want to obtain the longitude and latitude, but any conversion should stay inside the function and not modify the objects that were passed as inputs. Accessing `center.spherical.lat` and `center.spherical.lon` would avoid that problem.

eerovaher replied:
You could use `center.spherical.lat` and `center.spherical.lon` even if the `SkyCoord` already is in spherical coordinates, and then you wouldn't need `center.representation_component_names` at all.

eerovaher commented on astropy/visualization/wcsaxes/tests/test_images.py:
```diff
@@ -664,14 +664,23 @@ def test_patches_distortion(self, tmpdir):
+
+        r2 = SphericalCircle(SkyCoord(266.4 * u.deg, -29.1 * u.deg), 0.15 * u.degree,
+                             edgecolor='purple', facecolor='none',
+                             transform=ax.get_transform('fk5'))
```
There should also be a test for a `SkyCoord` that is not in spherical coordinates.

eerovaher commented on docs/changes/visualization/11790.feature.rst:
```diff
@@ -0,0 +1 @@
+Added a feature so that SphericalCircle will accept center parameter as an object.
```
It think it would be better if the change log entry would mention `SkyCoord` explicitly.

eerovaher commented on astropy/visualization/wcsaxes/patches.py:
```diff
@@ -2,9 +2,11 @@
 
 from astropy import units as u
+from astropy.coordinates import SkyCoord, SphericalRepresentation
 from astropy.coordinates.representation import UnitSphericalRepresentation
```
Importing `SphericalRepresentation` the way you did obviously works, but it looks a bit strange if `SphericalRepresentation` and `UnitSphericalRepresentation` are imported differently.

eerovaher commented on astropy/visualization/wcsaxes/patches.py:
```diff
@@ -81,9 +84,20 @@ class SphericalCircle(Polygon):
+                              'will be converted to Spherical Representation ',
+                              UserWarning)
+                center.representation_type = 'spherical'
+            attr_names = list(center.representation_component_names.keys())
```
Converting a dictionary into a list in Python only converts the keys, so 
```suggestion
            attr_names = list(center.representation_component_names)
```
would work equally well.

eerovaher commented on docs/visualization/wcsaxes/overlays.rst:
```diff
@@ -300,7 +300,7 @@ image to plot the contours for:
 *****************
 
-In the case where you are making a plot of a celestial image, and want to plot a circle that represents the area within a certain angle of a longitude/latitude, the `~matplotlib.patches.Circle` patch is not appropriate, since it will result in a distorted shape (because longitude is not the same as the angle on the sky). For this use case, you can instead use `~astropy.visualization.wcsaxes.Spher ...
+In the case where you are making a plot of a celestial image, and want to plot a circle that represents the area within a certain angle of a longitude/latitude, the `~matplotlib.patches.Circle` patch is not appropriate, since it will result in a distorted shape (because longitude is not the same as the angle on the sky). For this use case, you can instead use `~astropy.visualization.wcsaxes.Spher ...
```
The file https://github.com/astropy/astropy/blob/9819e926cb49fbc36019f5e641b563975e56d57c/docs/common_links.txt provides many useful aliases. It was added quite recently so a lot of the documentation does not make use of it yet, but I see no reason why new documentation shouldn't. Also this line looks very long, but if the rest of the file also contains overlong lines then it might be best to leave the line length as it is for consistency.

```suggestion
In the case where you are making a plot of a celestial image, and want to plot a circle that represents the area within a certain angle of a longitude/latitude, the `~matplotlib.patches.Circle` patch is not appropriate, since it will result in a distorted shape (because longitude is not the same as the angle on the sky). For this use case, you can instead use `~astropy.visualization.wcsaxes.SphericalCircle`, which takes a tuple of |Quantity| or a |SkyCoord| object as the input, and a |Quantity| as the radius:
```

AthulJoshi commented:
@eerovaher I think I've addressed everything. I guess all is good now. Sorry for the troubles (^_^").

eerovaher commented on astropy/visualization/wcsaxes/tests/test_images.py:
```diff
@@ -664,14 +665,28 @@ def test_patches_distortion(self, tmpdir):
+        # Test to verify that SphericalCircle works irrespective of whether
+        # the input(center) is a tuple or a SkyCoord object.
+        assert (r1.get_xy() == r2.get_xy()).all()
+        assert np.allclose(r1.get_xy(), r3.get_xy())
```
The test with a `SkyCoord` in Cartesian coordinates raises an `AstropyUserWarning`. You've told `pytest` to ignore this warning on line 632, but an even better approach would be to do the following instead:
```suggestion
        with pytest.warns(AstropyUserWarning, match=expected_message):
            assert np.allclose(r1.get_xy(), r3.get_xy())
```
where `expected_message` is the warning message you expect to receive. That way it becomes possible to test that the code only raises a warning where we expect it to, and the warning type and message are tested too.

AthulJoshi replied:
Ooh, I was looking for a good solution, sorry for not asking. I've made the changes. And now, should I squash all the commits?

eerovaher reviewed (changes requested):
The code itself looks good, but I do have a suggestion for one of the tests.

AthulJoshi commented:
@eerovaher All good? Sorry if you've already been notified.

eerovaher commented:
I think it would be a good idea to squash the commits now. @hamogu, what do you think?

hamogu commented:
Sure. All checks passed, changelog written, docs written, @eerovaher s points answered. That means we are very close, so I agree. Please do squash the commits to one or two. If you have not don so before, I recommend to make a backup first - I;ve screwed up squashing in the past myself!

AthulJoshi commented:
@eerovaher I've squashed the commits, The image tests failed? hmm. Did I mess up the squash

eerovaher commented:
I am not familiar with image tests, but there do seem to be problems running them at the moment, see issue #11841.

pllim commented:
Re: tests -- I restarted CircleCI in case failure was transient. I also approved Actions to run (something GitHub added to prevent Bitcoin miners).

hamogu commented:
@eerovaher knows a lot more about coordinates than I do and approved this already. Also looks good to me. So, I'll merge this!

Thanks a lot @AthulJoshi! I know it took longer than we both thought in the beginning, but I think it turned out well. I'm sure to use this in my projects as soon as I can. While it might seem like a little change here, I think for the user experience it's important to be able to use astropy objects where it makes sense. I clearly need a coordinate for plotting, so I should be able to pass in a coordinate. After all, what's the point of having a coordinate object, when I need to manually separate it in ra, dec to use it later anyway? 

Thank you for seeing this through and it would be great to see more code contributions from you ;-) Let me know if there is anything I can do to help.

AthulJoshi commented:
Thanks a lot @hamogu & @eerovaher. It's been a pleasure working on this. I got to learn a whole lot of things and this is my first contribution in a major open source community. Thanks a lot :D. This means so much for me. I'll try to find more issues and continue contributing.
